contents/security/Https.md

```diff
@@ -0,0 +1,59 @@
+
+There are mainly 3 encryption algorithms used in https, namely RSA, Diffie Hellman and Elliptic Curve Algorithm. They are more thoroughly explained in [Introduction to Cryptography](https://github.com/se-edu/learningresources "TODO: the link will be added after Dickson finishes on Cryptography") section. A simple understanding of why these algorithms prevent sniffers from decrypting the https packet by brute force is that the encryption process is of polynomial complexity and decryption by brute force is of exponential complexity. Therefore, when the encryption key size(number of bits in the key) is large enough(e.g. 2048 bits key for RSA), it becomes computationally impossible to decrypt the keys by brute force.
+
+Thus, by using https, we can be sure that even though our network packets are transmitted in an insecure network, sniffers cannot understand the content of our encrypted packets. 
+
+
+Besides providing secure network traffic, https also provides server validation through Certificate Authority(CA) architecture. A detailed explanation on CA is [here](https://www.globalsign.com/en-sg/ssl-information-center/what-are-certification-authorities-trust-hierarchies/) . In short, CA works by issuing the server a digital certificate that can only be produced by CA and client browsers can verify the server with digital certificate the server provides. 
+
+## How to set up https
+In order set up https on your server, you would need to have:
+1. A dedicated IP address for you server.
+2. Purchase a digital certificate from a CA.
+
+For example, one of the CA is [digicert](https://www.digicert.com) and they provide SSL [installation guide](https://www.digicert.com/ssl-certificate-installation.htm) for different platforms. Usually the SSL setup could be found on the CA that you purchased digital signature from.
+
+## Misuse of https
+
+Up to this moment, it seems that nothing could go wrong with https. However, in real life, there could be weakness on the implementation of https.
+
+1. In the implementation of https, the CA does not sign the digital certificate directly, rather it signes the fixed length hashed digest of the digital certificate for efficiency.This introduces the possibility of two different digital certificate with the same hashed digest. Thus, if the attackers manage to forge a fake digital certificate with the same hashed digest of another valid digital certificate, the browser would trust the attackers server and all the servers signed by the attackers. This loop hole occurs more often with SHA-1 hashing algorithm and SHA-1 is no longer used in https after 2016. A more detailed explanation of the problem is (here)[https://www.sott.net/article/275524-Why-HTTPS-and-SSL-are-not-as-secure-as-you-think]. However, the currently trusted SHA-2 might still be vulnerable to this attack.
```

Comment:
* signes -> signs
* Missing space: efficiency.This
* Add a link that explains hashes?
* "This loop hole occurs more often with SHA-1 hashing algorithm": don't think this has happened yet to real-world SHA1 certs - only reduced round SHA1 or theoretical scenarios. We're moving to sha256 just to be safe.
* The article's link has wrong markdown syntax. Also, the article is quite alarmist. Its true that SHA1 and SHA2 were invented by the NSA. But they are studied carefully by cryptographers. The weaknesses in SHA2 mentioned are only on reduced round versions. The NSA wouldn't need to resort to finding hash collisions - it can just force a US CA to generate any cert it wants.